superset-frontend/src/dashboard/components/DatasourceResultsPane/index.tsx

```diff
@@ -0,0 +1,232 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React, {
+  useState,
+  useEffect,
+  useMemo,
+  useCallback,
+  useRef,
+} from 'react';
+import {
+  BinaryQueryObjectFilterClause,
+  css,
+  Datasource,
+  ensureIsArray,
+  GenericDataType,
+  t,
+  useTheme,
+} from '@superset-ui/core';
+import Loading from 'src/components/Loading';
+import { EmptyStateMedium } from 'src/components/EmptyState';
+import TableView, { EmptyWrapperType } from 'src/components/TableView';
+import { useTableColumns } from 'src/explore/components/DataTableControl';
+import { getDatasourceSamples } from 'src/components/Chart/chartAction';
+import TableControls from './TableControls';
+
+const PAGE_SIZE = 50;
+
+export default function DatasourceResultsPane({
+  datasource,
+  initialFilters,
+}: {
+  datasource: Datasource;
+  initialFilters?: BinaryQueryObjectFilterClause[];
+}) {
+  const theme = useTheme();
+  const [pageIndex, setPageIndex] = useState(0);
+  const lastPageIndex = useRef(pageIndex);
+  const [filters, setFilters] = useState(initialFilters || []);
+  const [isLoading, setIsLoading] = useState(false);
+  const [responseError, setResponseError] = useState('');
+  const [resultsPages, setResultsPages] = useState<
```

Comment:
I understand your point - want to be able to cache the viewed samples, but this cache can become infinitely large, and the browser may crash. Do we consider not caching here?(Not working is better than being slow), or setting a length for the map.

Reply:
Hmm, good point.  I tried implementing a simple LRU with 5 elements but I'm worried that might not be enough to stop huge amounts of data from being cached.  Do you think what I have now helps enough or should we get rid of the cache altogether?

Reply:
Let's find a way to dynamically calculate the length of LRU. I open a [PR](https://github.com/apache/superset/pull/20841/files) for this. The samples_row_limit could be accessed by `common.conf.SAMPLES_ROW_LIMIT` in the Redux store. the length of the cache might be set to `floor(SAMPLES_ROW_LIMIT / PAGE_SIZE)` --- This means that the memory consumption of `DrillDetailPane` and `SamplesPane` is approximately the same.

Reply:
Ok, sounds good!